core/src/main/java/kafka/server/share/SharePartitionManager.java

```diff
@@ -617,22 +650,35 @@ private void maybeCompleteInitializationWithException(
             return;
         }
 
+        // Remove the partition from the cache as it's failed to initialize.
+        partitionCacheMap.remove(sharePartitionKey);
```

Comment:
1. We probably want to verify the leader epoch in SharePartition before removal to avoid a newly created SharePartition being removed by an old request.
2. We piggyback the removal in an error response. The downside is that there are quite a few error places to handle and it is reactive. An alternative is to have a partitionLeaderEpochChange listener. We can then proactively remove a SharePartition when the leader epoch changes.

Reply:
Both are great suggestions, do you think it would be right to have it in follow up PRs?